temporian/implementation/numpy/operators/assign.py

```diff
@@ -0,0 +1,97 @@
+
+"""Implementation for the Assign operator."""
+from typing import Dict, List
+
+import numpy as np
+from temporian.implementation.numpy.data.event import NumpyEvent
+from temporian.implementation.numpy.data.event import NumpyFeature
+from temporian.implementation.numpy.data.sampling import NumpySampling
+
+
+class NumpyAssignOperator:
+    def __call__(
+        self, assignee_event: NumpyEvent, assigned_event: NumpyEvent
+    ) -> Dict[str, NumpyEvent]:
+        """Assign features to an event.
+
+        Assignee and assigned must have same index names. Assigned cannot have more
+        than one row for a single index + timestamp occurence. Output event will
+        have same exact index and timestamps (sampling) as the assignee event.
+
+        Assignment is done by matching the timestamps and index of the assignee and assigned.
+        The assigned features will be appended to the assignee features in the matching
+        indexes. Index values in the assigned event missing in the assignee event will
+        be filled with NaNs.  Index values present in the assigned event missing in
+        the assignee event will be ignored.
+
+        Args:
+            assignee_event (NumpyEvent): event where features will be assigned to.
+            assigned_event (NumpyEvent): event with features to assign.
+
+        Returns:
+            NumpyEvent: a new event with the features assigned.
+
+        Raises:
+            ValueError: if assignee and assigned events have different indexes names.
+            ValueError: if assigned event has repeated timestamps for same index.
+
+        """
+        if assignee_event.sampling.names != assigned_event.sampling.names:
+            raise ValueError("Assign sequences must have the same index names.")
+
+        if assigned_event.sampling.has_repeated_timestamps:
+            raise ValueError(
+                "Assigned sequence cannot have repeated timestamps in the same"
+                " index."
+            )
+
+        output = NumpyEvent(
+            data=assignee_event.data.copy(), sampling=assignee_event.sampling
+        )
+
+        for index in assignee_event.data.keys():
+            # If index is in assigned append the features to the output event
+            if index in assigned_event.data.keys():
+                mask = (
+                    assignee_event.sampling.data[index][:, np.newaxis]
+                    == assigned_event.sampling.data[index][np.newaxis, :]
+                )
+                mask_i, mask_j = mask.nonzero()
+                for feature in assigned_event.data[index]:
+                    new_feature = NumpyFeature(
+                        name=feature.name,
+                        data=np.full(
+                            len(assignee_event.sampling.data[index]), np.nan
+                        ),
+                    )
+                    new_feature.data[mask_i] = feature.data[mask_j]
+                    output.data[index].append(new_feature)
```

Comment:
Nit: Can you cache the result of the query to "output.data[index]" over the "for feature in assigned_event.data[index]:" loop?